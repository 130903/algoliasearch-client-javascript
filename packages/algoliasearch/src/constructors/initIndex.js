// @flow

import * as indexMethods from '../methods/index/index.js';
import { createRequester } from 'algoliasearch-requester';
import attachParameters from './attachParameters.js';
import universalRequester from 'algoliasearch-http-requester';
import { AlgoliaError } from 'algoliasearch-errors';

import type { IndexMethods, AppId, ApiKey, IndexName } from 'algoliasearch';
import type {
  RequesterOptions,
  RequestMethod,
  HttpModule,
} from 'algoliasearch-requester';

export default function initIndex({
  appId,
  apiKey,
  indexName,
  requestOptions,
  requesterOptions,
  requester: extraRequester,
  httpRequester: extraHttpRequester,
}: {
  appId: AppId,
  apiKey: ApiKey,
  indexName: IndexName,
  requesterOptions?: RequesterOptions,
  requestOptions?: RequestOptions,
  requester?: RequestMethod,
  httpRequester?: HttpModule,
}): IndexMethods {
  /* eslint-disable prefer-rest-params */
  if (appId === undefined) {
    throw new AlgoliaError(`An appId is required. ${appId} was not valid.
    
    initIndex(${[...arguments].map(arg => JSON.stringify(arg)).join(',')})`);
  }
  if (apiKey === undefined) {
    throw new AlgoliaError(`An apiKey is required. ${apiKey} was not valid.

initIndex(${[...arguments].map(arg => JSON.stringify(arg)).join(',')})`);
  }
  if (indexName === undefined) {
    throw new AlgoliaError(`An indexName is required. ${indexName} was not valid.

initIndex(${[...arguments].map(arg => JSON.stringify(arg)).join(',')})`);
  }
  /* eslint-enable prefer-rest-params */

  const httpRequester =
    typeof extraHttpRequester === 'function'
      ? extraHttpRequester
      : universalRequester;

  const requester = extraRequester
    ? extraRequester
    : createRequester({
        appId,
        apiKey,
        httpRequester,
        requesterOptions,
        requestOptions,
      });

  // $FlowFixMe --> Flow doesn't get that the imports are augmented here
  return {
    ...attachParameters(indexMethods, { requester, indexName }),
    requester,
    httpRequester,
  };
}
