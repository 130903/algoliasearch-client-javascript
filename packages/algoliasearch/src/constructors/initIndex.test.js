// @flow
import initIndex from './initIndex';

const validParams = {
  appId: 'some_app',
  apiKey: 'some_key',
  indexName: 'some_index',
};

it('initIndex throws when it has too little parameters', () => {
  // $FlowIssue --> type disallows this
  expect(() => initIndex({})).toThrow();
  // $FlowIssue --> type disallows this
  expect(() => initIndex({ appId: '' })).toThrow();
  // $FlowIssue --> type disallows this
  expect(() => initIndex({ apiKey: '' })).toThrow();
  // $FlowIssue --> type disallows this
  expect(() => initIndex({ apiKey: '', appId: '' })).toThrow();

  expect(() => initIndex(validParams)).not.toThrow();
});

it('initIndex contains the correct methods', () => {
  const index = initIndex(validParams);
  expect(Object.keys(index)).toMatchSnapshot();
});
