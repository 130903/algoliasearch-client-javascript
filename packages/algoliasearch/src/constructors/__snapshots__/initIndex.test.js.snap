// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index.search() works üê£ 1`] = `
Object {
  "apiKey": "some_key",
  "appId": "some_app",
  "body": Object {
    "params": Object {
      "query": "hello world",
    },
  },
  "connectTimeout": 1000,
  "method": "POST",
  "requestOptions": undefined,
  "timeout": 2000,
  "url": Object {
    "hostname": "some_app-dsn.algolia.net",
    "pathname": "/1/indexes/some_index/queryundefined",
  },
}
`;

exports[`initIndex contains the correct methods 1`] = `
Array [
  "addApiKey",
  "addObject",
  "addObjects",
  "batch",
  "batchRules",
  "batchSynonyms",
  "browse",
  "browseAll",
  "browseFrom",
  "clearRules",
  "clearSynonyms",
  "deleteApiKey",
  "deleteBy",
  "deleteObject",
  "deleteObjects",
  "deleteRule",
  "deleteSynonym",
  "getApiKey",
  "getObject",
  "getObjects",
  "getRule",
  "getSettings",
  "getSynonym",
  "listApiKeys",
  "partialUpdateObject",
  "partialUpdateObjects",
  "saveObject",
  "saveObjects",
  "saveRule",
  "saveSynonym",
  "search",
  "searchForFacetValues",
  "searchRules",
  "searchSynonyms",
  "setSettings",
  "updateApiKey",
  "waitTask",
  "requester",
  "httpRequester",
]
`;

exports[`initIndex throws when it has too little parameters 1`] = `
"An apiKey is required. undefined was not valid.

initIndex({\\"appId\\":\\"\\"})"
`;

exports[`initIndex throws when it has too little parameters 2`] = `
"An appId is required. undefined was not valid.
    
    initIndex({\\"apiKey\\":\\"\\"})"
`;

exports[`initIndex throws when it has too little parameters 3`] = `
"An indexName is required. undefined was not valid.

initIndex({\\"apiKey\\":\\"\\",\\"appId\\":\\"\\"})"
`;
